python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,232 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+import copy
+import cudf
+import ctypes
+import math
+import numpy as np
+import warnings
+
+from numba import cuda
+
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+from libc.stdlib cimport calloc, malloc, free
+
+from cuml.common.base import Base
+from cuml.common.handle cimport cumlHandle
+from cuml.utils import get_cudf_column_ptr, get_dev_array_ptr, \
+    input_to_dev_array, zeros
+cimport cuml.common.handle
+cimport cuml.common.cuda
+
+cdef extern from "treelite/c_api.h":
+    ctypedef void* ModelHandle
+
+cdef extern from "fil/fil.h" namespace "ML::fil":
+
+    cdef enum algo_t:
+        NAIVE,
+        TREE_REORG,
+        BATCH_TREE_REORG
+
+    cdef enum output_t:
+        RAW,
+        PROB,
+        CLASS
+
+    cdef struct dense_node_t:
+        float val
+        int bits
+
+    cdef struct forest:
+        pass
+
+    ctypedef forest* forest_t
+
+    cdef struct treelite_params_t:
+        algo_t algo
+        bool output_class
+        float threshold
+
+    cdef void free(cumlHandle& handle,
+                   forest_t)
+
+    cdef void predict(cumlHandle& handle,
+                      forest_t,
+                      float*,
+                      float*,
+                      size_t)
+
+    cdef forest_t from_treelite(cumlHandle& handle,
+                                forest_t*,
+                                ModelHandle,
+                                treelite_params_t*)
+
+cdef class FIL_impl():
+
+    cpdef object handle
+    cdef forest_t forest_data
+    cdef object algo
+    cdef object threshold
+
+    def __cinit__(self,
+                  algo=0,
+                  threshold=0.0,
+                  handle=None):
+        self.algo = algo
+        self.threshold = threshold
+        self.handle = handle
+
+    def predict(self, X, preds=None):
+        cdef uintptr_t X_ptr
+        X_m, X_ptr, n_rows, _, X_dtype = \
+            input_to_dev_array(X, order='C')
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        if preds is None:
+            preds = cudf.Series(zeros(n_rows, dtype=np.float32))
```

Comment:
A numba array might be more lightweight, but I'm not an expert here, so it's up to you.